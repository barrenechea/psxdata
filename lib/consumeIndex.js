import LANGUAGE_MAP from "./languages.js";

function findTextNodes(nodeList) {
  return Array.from(nodeList).filter((node) => node.nodeType === 3);
}

export default function consumeIndex(document) {
  return Array.from(document.querySelectorAll(".sectiontable tr"))
    .filter((row) => row.childElementCount)
    .map((row) => {
      const data = {};

      // Get IDs from list
      data.id = findTextNodes(row.querySelector(".col2, .col6").childNodes)
        .map((node) => node.textContent.trim())
        .filter((text) => text);

      // Infer disc count from IDs
      const discs = data.id.length;

      // If only one ID, switch to a string
      if (data.id.length === 1) {
        data.id = data.id.shift();
      }

      // Parse out just the title (ignore disc count)
      const titleColumn = row.querySelector(".col3, .col7");

      data.title = titleColumn.firstChild.textContent
        .replace(/\[\s*\d DISCS\s*\]\]?/g, "")
        .replace(/^[\s-]+|[\s-]+$/g, "");

      // Parse out any extra title info, such as disc titles
      const extraHeadings = Array.from(
        row.querySelectorAll(".col3 > span > u, .col7 > span > u")
      );

      if (extraHeadings.length) {
        extraHeadings.forEach((heading) => {
          const headingTitle = heading.textContent
            .toLowerCase()
            .replace(/^\s*|[\:\s]*$/g, "");

          const headingDetailEntries = findTextNodes(
            heading.nextElementSibling.childNodes
          ).map((text) => text.textContent.replace(/[\s\n]+/g, " ").trim());

          if (headingDetailEntries.length === 1) {
            data[headingTitle] = headingDetailEntries.shift();
          } else if (headingDetailEntries.length > 1) {
            data[headingTitle] = headingDetailEntries;
          }
        });
      }

      data.discs = discs;

      // Parse language list
      const languages = row
        .querySelector(".col4, .col8")
        .textContent.match(/\w+/g);
      if (languages) {
        data.languages = languages.map(
          (language) => LANGUAGE_MAP[language.toLowerCase()] || language
        );
      }

      // Parse PSXDataCenter detail link
      const link = row.querySelector(".col1 a[href], .col5 a[href]");
      if (link) {
        data.link = link.href;
      }

      return data;
    });
}
