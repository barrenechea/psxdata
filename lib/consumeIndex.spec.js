import path from "path";
import { fileURLToPath } from "url";
import { JSDOM } from "jsdom";
import consumeIndex from "./consumeIndex.js";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

describe("consumeIndex", () => {
  it("produces expected output under JSDOM", async () => {
    const dom = await JSDOM.fromFile(
      path.join(__dirname, "__fixtures__", "ps1.html"),
      { url: "https://test-psxdata" }
    );

    expect(consumeIndex(dom.window.document)).toMatchSnapshot();
  });
});
