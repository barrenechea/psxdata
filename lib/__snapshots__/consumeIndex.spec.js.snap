// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`consumeIndex produces expected output under JSDOM 1`] = `
[
  {
    "discs": 6,
    "id": [
      "SLUS-01224",
      "SLUS-01252",
      "SLUS-01253",
      "SLUS-01254",
      "SLUS-01255",
      "SLUS-01256",
    ],
    "includes": [
      "Arc The Lad - Arc The Lad II - Arc The Lad III (2 Discs)",
      "Arc The Lad - Monster Tournament - Battle Arena & Arc The Lad - The Making Of Disc",
    ],
    "languages": [
      "en",
    ],
    "link": "https://test-psxdata/games/U/A/SLUS-01224.html",
    "title": "ARC THE LAD COLLECTION",
  },
  {
    "discs": 2,
    "id": [
      "SLES-00530",
      "SLES-10530",
    ],
    "includes": "GDI Disc + NOD Disc",
    "languages": [
      "en",
    ],
    "link": "https://test-psxdata/games/P/C/SLES-00530.html",
    "title": "COMMAND & CONQUER",
  },
  {
    "discs": 2,
    "id": [
      "SLES-00531",
      "SLES-10531",
    ],
    "includes": "GDI Disc + NOD Disc",
    "languages": [
      "fr",
    ],
    "link": "https://test-psxdata/games/P/C/SLES-00531.html",
    "title": "COMMAND & CONQUER",
  },
  {
    "discs": 1,
    "id": "SLES-00927",
    "languages": [
      "en",
      "fr",
      "de",
      "es",
    ],
    "link": "https://test-psxdata/games/P/C/SLES-00927.html",
    "title": "CONSTRUCTOR",
  },
  {
    "discs": 1,
    "id": "SCES-00344",
    "languages": [
      "en",
    ],
    "link": "https://test-psxdata/games/P/C/SCES-00344.html",
    "title": "CRASH BANDICOOT",
  },
  {
    "discs": 1,
    "id": "SCES-00967",
    "languages": [
      "en",
      "fr",
      "de",
      "it",
      "es",
    ],
    "link": "https://test-psxdata/games/P/C/SCES-00967.html",
    "title": "CRASH BANDICOOT 2- CORTEX STRIKES BACK",
  },
  {
    "discs": 1,
    "id": "SCES-01420",
    "languages": [
      "en",
      "fr",
      "de",
      "it",
      "es",
    ],
    "link": "https://test-psxdata/games/P/C/SCES-01420.html",
    "title": "CRASH BANDICOOT 3 - WARPED",
  },
  {
    "discs": 1,
    "id": "SCES-02834",
    "languages": [
      "en",
      "fr",
      "de",
      "it",
      "es",
    ],
    "link": "https://test-psxdata/games/P/C/SCES-02834.html",
    "title": "CRASH BASH",
  },
  {
    "discs": 1,
    "id": "SCES-02105",
    "languages": [
      "en",
      "fr",
      "de",
      "it",
      "es",
    ],
    "link": "https://test-psxdata/games/P/C/SCES-02105.html",
    "title": "CRASH TEAM RACING",
  },
  {
    "discs": 1,
    "id": "SCES-00984",
    "languages": [
      "en",
      "fr",
      "de",
      "it",
      "es",
    ],
    "link": "https://test-psxdata/games/P/G/SCES-00984.html",
    "title": "GRAN TURISMO - THE REAL DRIVING SIMULATOR",
  },
  {
    "discs": 2,
    "id": [
      "SCES-02380",
      "SCES-12380",
    ],
    "includes": "Arcade Disc + Simulation Disc",
    "languages": [
      "en",
      "fr",
      "de",
      "it",
      "es",
    ],
    "link": "https://test-psxdata/games/P/G/SCES-02380.html",
    "title": "GRAN TURISMO 2 - THE REAL DRIVING SIMULATOR",
  },
  {
    "discs": 1,
    "id": "SLES-00032",
    "languages": [
      "en",
      "fr",
      "de",
      "it",
    ],
    "link": "https://test-psxdata/games/P/G/SLES-00032.html",
    "title": "GRAND THEFT AUTO",
  },
  {
    "discs": 3,
    "id": [
      "SLES-00032",
      "SLES-03389",
      "SLES-01404",
    ],
    "includes": "Grand Theft Auto Disc + Mission Pack #1 - London 1969 + Grand Theft Auto 2 Disc",
    "languages": [
      "en",
      "fr",
      "de",
      "it",
      "es",
    ],
    "link": "https://test-psxdata/games/P/G/SLES-00032CE.html",
    "title": "GRAND THEFT AUTO [COLLECTOR'S EDITION SET]",
  },
  {
    "discs": 1,
    "id": "SLES-03389",
    "languages": [
      "en",
      "fr",
      "de",
      "it",
      "es",
    ],
    "link": "https://test-psxdata/games/P/G/SLES-03389.html",
    "title": "GRAND THEFT AUTO - LONDON [SPECIAL EDITION]",
  },
  {
    "discs": 1,
    "id": "SLES-01714",
    "languages": [
      "en",
      "fr",
      "de",
      "it",
      "es",
    ],
    "link": "https://test-psxdata/games/P/G/SLES-01714.html",
    "title": "GRAND THEFT AUTO - MISSION PACK #1 - LONDON 1969",
  },
  {
    "discs": 1,
    "id": "SLES-01404",
    "languages": [
      "en",
      "fr",
      "de",
      "it",
      "es",
    ],
    "link": "https://test-psxdata/games/P/G/SLES-01404.html",
    "title": "GRAND THEFT AUTO 2",
  },
  {
    "discs": 1,
    "id": "SLES-02458",
    "languages": [
      "en",
      "fr",
      "de",
      "it",
      "es",
    ],
    "link": "https://test-psxdata/games/P/G/SLES-02458.html",
    "title": "GRAND THEFT AUTO 2",
  },
]
`;
